Add generic observation processes which combine the convolution with the noise model.

cdc-mitzimorris opened the pull request:
This PR adds work that was done in https://github.com/cdcent/cfa-pyrenew-hierarchical/pull/4   to PyRenew.

It adds the base observation process class, concrete implementations for Count processes and the abstract base class for Measurement processes,  together with unit tests and two new tutorials for count and measurement observation processes respectively.

Once this PR and the work done in https://github.com/cdcent/cfa-pyrenew-hierarchical/pull/5 have been added to PyRenew, subsequent PRs will deprecate unused features and harmonize the documentation and tutorials.

cdc-mitzimorris commented:
thanks @dylanhmorris  - all suggested changes made.

dylanhmorris commented on pyrenew/observation/noise.py:
```diff
@@ -0,0 +1,372 @@
+Count Noise
+-----------
+- ``PoissonNoise``: Equidispersed counts (variance = mean). No parameters.
+- ``NegativeBinomialNoise``: Overdispersed counts (variance > mean).
```
```suggestion
- ``NegativeBinomialNoise``: Overdispersed counts relative to Poisson (variance > mean).
```

dylanhmorris commented on pyrenew/observation/noise.py:
```diff
@@ -0,0 +1,372 @@
+_EPSILON = 1e-10
+
+
+class CountNoise(ABC):
```
Why are `CountNoise` and `MeasurementNoise` separate abstract base classes and not subclasses of an overall abstract `Noise` class

dylanhmorris commented on pyrenew/observation/noise.py:
- Any reason not to make `noise` its own module rather than a sub-module of observation, even if in practice we'll main use it for observations? @damonbayer thoughts/preferences?

damonbayer replied:
I don't have a strong opinion, but lean toward making `noise` its own module.

dylanhmorris commented on pyrenew/observation/count_observations.py:
```diff
@@ -0,0 +1,348 @@
+        infections: ArrayLike,
+        subpop_indices: ArrayLike,
+        times: ArrayLike,
+        counts: ArrayLike | None = None,
```
think we should name this `obs` throughout.

dylanhmorris commented on pyrenew/observation/base.py:
```diff
@@ -0,0 +1,334 @@
+        pass  # pragma: no cover
+
+    @abstractmethod
+    def sample(self, **kwargs) -> ArrayLike:
```
Should have an `obs` kwarg specified to standardize the interface (per examples below).

dylanhmorris commented on pyrenew/randomvariable/hierarchical.py:
```diff
@@ -0,0 +1,336 @@
+            group_sd = numpyro.deterministic(
+                self.name,
+                jnp.maximum(raw_sd, self.sd_min),
+            )
```
Probably simpler and cleaner to Numpyro's truncated distribution API https://num.pyro.ai/en/stable/tutorials/truncated_distributions.html

dylanhmorris commented on pyrenew/observation/noise.py:
```diff
@@ -0,0 +1,372 @@
+            If sensor_sd samples non-positive values.
+        """
+        sensor_mode = self.sensor_mode_prior_rv.sample(n_groups=n_sensors)
+        sensor_sd = self.sensor_sd_prior_rv.sample(n_groups=n_sensors)
```
Why not plate here? Then `sensor_mode_prior_rv` and `sensor_sd_prior_rv` don't need the n_groups interface.

dylanhmorris reviewed (changes requested):
Thanks @cdc-mitzimorris! A high-level suggested changes before re-review

- Change "expected" --> "predicted" throughout, as nothing in the design requires or enforces that the prediction be the expected value, though in many concrete cases it is, and that's desirable.
- Throughout, avoid custom names for the observed quantity, if any, to the `sample()` call. Instead use `obs`, for API mirroring of `numpyro.sample()`.
- Revise the measurements tutorial to be stricter about what is specific to the wastewater example and what is general to delay-ascertained measurements.
- Remove the provided helper classes for sampling priors (i.e. `pyrenew/randomvariable/hierarchical.py` from this PR. That will keep it more narrowly focused on observation processes. 


RE: the helper classes for sampling priors: I am open to the possibility that we could want those helper classes in Pyrenew, but as I mention in my line-by-line comments, I think it requires a broader rethink of how we do distributional variables, etc. Would also appreciate thoughts from @damonbayer on that point.

cdc-mitzimorris commented:
ready for re-review - high-level concerns addressed:

- changed `expected` to `predicted`
- standardized calls to `sample` w/ arg `obs`
- revised measurement observations tutorial
- removed helper class for hierarchical models; use NumPyro's TruncatedNormalDistribution directly.

damonbayer commented on docs/tutorials/observation_processes_counts.qmd:
````diff
@@ -0,0 +1,667 @@
+```
+</details>
+
+To see the reduction in noise, it is ncessary to keep the y-axis on the same scale as in the previous plot.
````
```suggestion
To see the reduction in noise, it is necessary to keep the y-axis on the same scale as in the previous plot.
```

dylanhmorris commented on docs/tutorials/observation_processes_counts.qmd:
```diff
@@ -63,47 +83,64 @@ In this example, we use fixed parameter values for illustration; in practice, th
 For example, if `hosp_delay_pmf[5] = 0.2`, then 20% of infections that result in hospitalization will appear as hospital admissions 5 days after infection.
 
-We load a delay distribution from PyRenew's datasets:
+We load a delay distribution from PyRenew's datasets which peaks around day 8-9 post-infection, compute summary statistics, and plot it.
```
```suggestion
We load a delay distribution from PyRenew's example datasets, compute summary statistics, and plot it. The distribution peaks around day 8-9 post-infection.
```

dylanhmorris commented on docs/tutorials/observation_processes_measurements.qmd:
```diff
@@ -14,58 +23,174 @@ import jax.numpy as jnp
+
+### The noise model
+
+Measurement data typically exhibits **sensor-level variability**: different instruments, labs, or sampling locations have systematic biases and different precision levels.
```
```suggestion
Measurement data typically exhibits **sensor-level variability**: different instruments, labs, or sampling locations may have systematic biases and different precision levels.
```

dylanhmorris commented on pyrenew/observation/noise.py:
```diff
@@ -0,0 +1,372 @@
+
+from pyrenew.metaclass import RandomVariable
+
+_EPSILON = 1e-10
```
Still need to do this.

dylanhmorris commented on pyrenew/observation/measurements.py:
```diff
@@ -0,0 +1,144 @@
+        """
+        predicted_values = self._predicted_obs(infections)
+
+        self._deterministic("predicted_log_conc", predicted_values)
```
Site name should be configurable and/or dynamically generated based on a higher-level configurable name.

dylanhmorris reviewed (changes requested):
Thanks @cdc-mitzimorris! A few things to address and then I can re-review.

dylanhmorris commented on pyrenew/observation/base.py:
```diff
@@ -262,14 +259,36 @@ def _deterministic(self, name: str, value: ArrayLike) -> None:
+            Full site name with ``self.name`` prefix.
+            For example, suffix="obs" returns "{name}_obs".
+        """
+        return f"{self.name}_{suffix}"
```
Consider using https://num.pyro.ai/en/latest/handlers.html#scope to accomplish this (your call on whether to do now or in a separate PR

dylanhmorris commented on test/test_observation_measurements.py:
```diff
@@ -284,7 +288,7 @@ def test_predicted_obs_stored(self):
             ).get_trace()
 
-        assert "predicted_log_conc" in trace
+        assert "test_predicted" in trace
```
good to have a test of the programmatic naming. I believe we want one for obs too, no?

dylanhmorris reviewed (commented):
Thanks @cdc-mitzimorris! Just a couple remaining questions.

dylanhmorris commented on pyrenew/observation/base.py:
```diff
@@ -0,0 +1,356 @@
+            Value to track. Can be any shape.
+        """
+        site_name = f"{self.name}_{suffix}"
+        numpyro.deterministic(site_name, value)
```
Use the helper function
```suggestion
        numpyro.deterministic(self._sample_site_name(suffix), value)
```

dylanhmorris commented on pyrenew/observation/base.py:
```diff
@@ -0,0 +1,356 @@
+            )
+
+        if jnp.any(pmf < 0):
+            raise ValueError(f"{param_name} must have non-negative values")
```
Should land this PR and then address, but this duplicates `distutil.validate_discrete_dist_vector`.

That said, I prefer the implementation here, so this should replace that.

We could also (in #645) consider having discrete PMFs as a strong abstraction. We could then require certain RVs to be of class `PMF` and deferring the validation to those RVs' constructors. Given how fundamental discrete PMFs are to discrete-time renewal processes I favor this.

cdc-mitzimorris replied:
noted.  marking as resolved.

dylanhmorris commented on pyrenew/observation/base.py:
```diff
@@ -0,0 +1,356 @@
+            Equal to ``len(pmf) - 1``.
+        """
+        pmf = self.temporal_pmf_rv()
+        return int(len(pmf) - 1)
```
Not used anywhere in the class. Would be better to implement in terms of a PMF abstraction (could then also apply to computing the offset in `compute_delay_ascertained_incidence`. I think refrain from implementing for now.
```suggestion
```

dylanhmorris commented on pyrenew/observation/count_observations.py:
```diff
@@ -0,0 +1,314 @@
+
+        predicted_counts = jax.vmap(convolve_col, in_axes=1, out_axes=1)(infections)
+
+        return predicted_counts[:, 0] if is_1d else predicted_counts
```
If there's going to be a switch for 1d input, I think it's cleaner and clearer to `vmap` only in the >2d case, and not pad the input only to unpad the output
```suggestion
        is_1d = infections.ndim == 1
        if is_1d:
            predicted_counts = self._convolve_with_alignment(infections, delay_pmf, ascertainment_rate)[0]
        else: 
            predicted_counts = jax.vmap(lambda col: self._convolve_with_alignment(col, delay_pmf, ascertainment_rate)[0], in_axes=1, out_axes=1)(infections)
        return predicted_counts
```

dylanhmorris commented on pyrenew/observation/count_observations.py:
```diff
@@ -0,0 +1,314 @@
+            Subpopulation index for each observation (0-indexed).
+            Shape: (n_obs,)
+        times : ArrayLike
+            Day index for each observation (0-indexed).
```
```suggestion
            Day index for each observation (0-indexed) relative to the infections vector.
```

dylanhmorris commented on test/test_observation_counts.py:
```diff
@@ -0,0 +1,564 @@
+
+        # Timeline alignment: output length equals input length
+        assert result.observed.shape[0] == len(infections)
+        # First len(delay_pmf)-1 days are NaN (appear as -1 after NegativeBinomial sampling)
```
```suggestion
```

dylanhmorris commented on test/test_observation_counts.py:
Not for this PR: a test of `Counts` sampling with `obs=None` where observations are dense and the delay PMF is of length > 1 would have revealed the issue with NaN padding and -1s. Should add.

dylanhmorris reviewed (approved):
Thanks @cdc-mitzimorris! Left suggestions that should fix the tutorial rendering. Apologies for failing to make those along with the test changes in my last review. Please merge when ready!
